python/app/input_widget.py

```diff
@@ -227,20 +228,364 @@ def keyPressEvent(self, event):
 
         :param event: key press event object.
         """
-
-        if event.modifiers() & QtCore.Qt.ShiftModifier and event.key() in [
+        if event.key() in [
             QtCore.Qt.Key_Enter,
             QtCore.Qt.Key_Return,
         ]:
-            self.insertPlainText("\n")
+            self.add_new_line()
+            event.accept()
+        elif event.key() == QtCore.Qt.Key_Backspace:
+            # Attempt to remove the indentation if the cursor is within the indentation section of the line
+            # other wise fall back to default behaviour.
+            if self.remove_character_indentation():
+                event.accept()
+            else:
+                super(PythonInputWidget, self).keyPressEvent(event)
+        elif event.key() == QtCore.Qt.Key_Slash:
+            self.block_comment_selection()
+            event.accept()
+        elif event.key() == QtCore.Qt.Key_Backtab:
+            self.unindent()
             event.accept()
         elif event.key() == QtCore.Qt.Key_Tab:
             # intercept the tab key and insert 4 spaces
-            self.insertPlainText("    ")
+            self.indent()
             event.accept()
         else:
             super(PythonInputWidget, self).keyPressEvent(event)
 
+    def remove_character_indentation(self):
+        """
+        Attempts to remove a single indentation block if there is no selection and the cursor is in the indentation
+        part of the line. It returns True if we unindented and False if we didn't.
+        :return: bool
+        """
+        cur = self.textCursor()
+        if not cur.hasSelection():
+            user_cur_pos = cur.positionInBlock()
+            line = cur.block().text()
+
+            # Now separate out the indentation from the rest of the line and figure out how many spaces it is.
+            indentation, rest_of_line = self._split_indentation(line)
+
+            # Te current user cursor position is within the indentation.
+            if user_cur_pos <= len(indentation) and len(indentation) != 0:
+                self.unindent()
+                return True
+        return False
+
+    def add_new_line(self):
+        """
+        Adds a new line from the cursor position.
+        The new line will be indented at the same level as the previous line.
+        :return: None
+        """
+        cur = self.textCursor()
+
+        # Start the undo block
```

Comment:
Could you comment how the undoBlock helps here?

Reply:
Oh, and it says undo block but the code says editBlock actually!

Reply:
I've updated the comment. `beginEditBlock` will create a start point for recording the changes, the `endEditBlock` will stop it, and then if you undo, it will undo everything made between those two points as one change rather than multiple.

Reply:
It just dawned on me that if there is an exception raised in your code you're likely to leave the widget in a weird state. Should we put everything inside a 
```
begitEditBlock
try:
    ...
    should_undo = Fase
except:
    should_undo = True
finally:
    endEditBlock

if should_undo:
    undo
```

or something like it?

Reply:
It's a good point, I'll have a look!

Reply:
OK, I've added a new method that can be used as a decorator. It handles exactly what you suggested, and I've added the decorator to the methods that are altering the text.
good catch!